src/pkg/artifact/dao/dao.go

```diff
@@ -0,0 +1,114 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dao
+
+import (
+	"context"
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/pkg/q"
+)
+
+// DAO is the data access object interface for artifact
+type DAO interface {
+	// Count returns the total count of artifacts according to the query
+	Count(ctx context.Context, query *q.Query) (total int64, err error)
+	// List artifacts according to the query
+	List(ctx context.Context, query *q.Query) (artifacts []*Artifact, err error)
+	// Get the artifact specified by ID
+	Get(ctx context.Context, id int64) (*Artifact, error)
+	// Create the artifact
+	Create(ctx context.Context, artifact *Artifact) (id int64, err error)
+	// Delete the artifact specified by ID
+	Delete(ctx context.Context, id int64) (err error)
+	// Update updates the artifact. Only the properties specified by "props" will be updated if it is set
+	Update(ctx context.Context, artifact *Artifact, props ...string) (err error)
+	// CreateReference creates the artifact reference
+	CreateReference(ctx context.Context, reference *ArtifactReference) (id int64, err error)
+	// ListReferences lists the artifact references according to the query
+	ListReferences(ctx context.Context, query *q.Query) (references []*ArtifactReference, err error)
+	// DeleteReferences deletes the references referenced by the artifact specified by parent ID
+	DeleteReferences(ctx context.Context, parentID int64) (err error)
+}
+
+// New returns an instance of the default DAO
+func New() DAO {
+	return &dao{}
+}
+
+type dao struct{}
```

Comment:
Generally speaking, shall we add a bit more logic here such as sanity check for parms or shield user from transforming int64 -> int